### PR TITLE
VIS Diagrams in API & UI

### PR DESCRIPTION
**Pull request by sunu**
Refs #886, also alephdata/vislib#120.

### CODE REVIEW

File: aleph/model/diagram.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+
+log = logging.getLogger(__name__)
+
+
+class Diagram(db.Model, SoftDeleteModel):
+    """A mapping to load entities from a table"""
+    __tablename__ = 'diagram'
+
+    id = db.Column(db.Integer, primary_key=True)
```
Comment (pudo) [inline aleph/model/diagram.py]: Can we also have a `label`?

File: aleph/validation/schema/diagram.yml [hunk 2]
```diff
@@ -0,0 +1,115 @@
+DiagramCreate:
+  type: object
+  required: ["data"]
+  properties:
+    data:
+      type: object
+      allOf:
+      - $ref: '#/components/schemas/DatedModel'
```
Comment (pudo) [inline aleph/validation/schema/diagram.yml]: I think that wants to move up one level.

File: aleph/logic/diagram.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        proxy = obj.to_proxy()
+        obj.delete()
+        index.delete_entity(entity_id)
+        refresh_entity(proxy)
+
+
+def delete_diagram(diagram):
+    existing_ids = diagram.entity_ids
+    # Clean up created entities
+    for ent_id in existing_ids:
```
Comment (pudo) [inline aleph/logic/diagram.py]: I reckon this should either be a flag, or maybe not even exist at all :/

**Comment by sunu** [comment]
> Question: what's the up- and down-side of making `entities` an ARRAY on the table? I assume we can query within the data blob, too, but this might be more convenient.
> 
I have this property function to access the entity ids in a diagram: https://github.com/alephdata/aleph/blob/0f7784de36b4a60ba45e8958d62bb4b0605dfc40/aleph/model/diagram.py#L64 My reasoning behind storing all the data in one place is just to keep the data as close to the original VIS format as possible. I don't have any strong reasons for or against it.

> There's something icky about letting VIS choose database entity IDs. These need to be globally unique (they're the table's primary key), and we could provoke conflicts here easily (i.e. it's an attack vector). I'm almost curious if VIS should negotiate this with Aleph via a hook in VIS' entity creation routines...

Right now, we are throwing away the id given by VIS and making our own id: https://github.com/alephdata/aleph/blob/0f7784de36b4a60ba45e8958d62bb4b0605dfc40/aleph/model/entity.py#L81 So the blob returned by the backend after creation has completely different ids that the input blob.

File: aleph/validation/schema/diagram.yml (hunk 2 above)
Comment (sunu) [reply aleph/validation/schema/diagram.yml]: It's a type. Should be `DiagramData` instead.

File: aleph/model/diagram.py (hunk 1 above)
Comment (sunu) [reply aleph/model/diagram.py]: Yes, makes sense

File: aleph/logic/diagram.py (hunk 3 above)
Comment (sunu) [reply aleph/logic/diagram.py]: Ok, I'll make it a flag for now. If we don't have it at all, is there any other user-facing way to delete stray entities later?

File: aleph/validation/schema/diagram.yml [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+              position:
+                type: object
+                properties:
+                  x:
+                    type: number
+                  y:
+                    type: number
+              entityId:
+                type: string
+        edged:
```
Comment (kjacks) [inline aleph/validation/schema/diagram.yml]: edged => edges

File: aleph/validation/schema/diagram.yml [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                properties:
+                  x:
+                    type: number
+                  y:
+                    type: number
+        groupings:
+          type: array
+        selection:
+          type: array
+    viewport:
```
Comment (kjacks) [inline aleph/validation/schema/diagram.yml]: @sunu I'm thinking to actually only store the layout information, not the viewport.  But curious if you have any thoughts on this.  The viewport is only storing the bounding box/zoom level so would only be necessary to store if we wanted to show a user the same view of the diagram that they had left previously

Comment (sunu) [reply aleph/validation/schema/diagram.yml]: In that case, omitting viewport seems reasonable. Do you think there's anything else we should not store? May be selection (I assume that's storing the current selection)?

File: aleph/model/diagram.py (hunk 1 above)
Comment (kjacks) [reply aleph/model/diagram.py]: And maybe a description?

**Pull request closed without merging**